### ISSUE TITLE
Links to protocol-relative URLs are not marked rel=nofollow

### ISSUE THREAD

**Issue by iliana** [rehypejs/rehype-external-links#8]
### Initial checklist

- [X] I read the [support docs](https://github.com/rehypejs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/rehypejs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/rehypejs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Arehypejs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

1.0.0-2.0.0

### Link to runnable example

https://codesandbox.io/s/relaxed-sound-irixmt?file=/src/index.js

### Steps to reproduce

Based on the example in the README:

```js
import {unified} from 'unified'
import rehypeParse from 'rehype-parse'
import rehypeExternalLinks from 'rehype-external-links'
import rehypeStringify from 'rehype-stringify'

const file = await unified()
  .use(rehypeParse)
  .use(rehypeExternalLinks, {rel: ['nofollow']})
  .use(rehypeStringify)
  .process('<a href="//github.com/rehypejs/rehype">rehype</a>')

console.log(String(file))
```

This example uses a protocol-relative URL, which is [widely considered an anti-pattern in actual development](https://www.paulirish.com/2010/the-protocol-relative-url/), but is still supported by browsers.

### Expected behavior

Protocol-relative URLs include the domain name, and thus on web sites are just like an external http/https URL. Thus we expect:

```html
<html><head></head><body><a href="//github.com/rehypejs/rehype" rel="nofollow">rehype</a></body></html>
```

### Actual behavior

Because there is no protocol, [this particular line](https://github.com/rehypejs/rehype-external-links/blob/eafd7c2e02332c92d87399f809c779d995494378/index.js#L93) determines the protocol for this URL is `//github.com/rehypejs/rehyp` which is not in any reasonably-configured protocol list, so rel="nofollow" is not added.

```html
<html><head></head><body><a href="//github.com/rehypejs/rehype">rehype</a></body></html>
```

This behavior was introduced in, best I can tell, https://github.com/remarkjs/remark-external-links/pull/10, which added this option because of weird interactions between `mailto:` links and `target="_blank"`, the latter of which has since been removed from this codebase.

### Runtime

Node v16

### Package manager

npm 8

### OS

macOS

**Comment by wooorm** [comment]
Heya!

* Perhaps the “current” (assumed) protocol might have to be passed, defaulting to say `http`? This might be useful to have because we have [a `protocols` option](https://github.com/rehypejs/rehype-external-links#optionsprotocols).
* Would it be needed to support this on, say, an `ftp`, `gemini` or something? Not all protocols need slashes (e.g., `tel`, `mailto`), what to do there?
* [The code here](https://github.com/micromark/micromark/blob/0505869674f968735bdaa14198fd7ac230b205d2/packages/micromark-util-sanitize-uri/dev/index.js#L32-L51) (different project) can properly find if a dangerous protocol is used, assuming the page will be on a safe protocol instead. Something like that might be useful here but is not enough, for your slashes example

---

Alternatively, perhaps this project would be improved if we a) pass the URL where the page will hosted, b) run `new URL(a.href, baseUrl)`, c) check if the domains are equal or different.
This has the downside of potentially broken URLs either crashing or alternatively always marked as external (?), but the upside of full URLs with protocols or slashes that go to my own domain being marked as safe

**Comment by iliana** [comment]
I wish `new URL` didn't require a base URL to be a useful, robust URL parser. :( Requiring that to be set by the caller to get this check would probably result in nobody getting this check.

Probably the safest thing to do is assume that any href starting with `//` is an external URL ([RFC 3986 § 3](https://datatracker.ietf.org/doc/html/rfc3986#section-3); `When authority is not present, the path cannot begin with two slash characters ("//").`).

Given that protocol-relative URLs are frowned up on in web dev, I think the most likely use of a protocol-relative URL in the wild is someone who is trying to work around this package's checks (and I say this as someone who reported this bug because I was able to work around it in the wild).

**Comment by wooorm** [comment]
> Requiring that to be set by the caller to get this check would probably result in nobody getting this check.

This sounds a bit “adamant”. Why so certain? We could even require it and throw if it isn’t configured. I think it’s quite an improvement to parse “full” URLs as well.

Some plugins need it too, and they currently support:

```js
const file = await read('example.md')

// Define where the generated file will be available.
file.data.meta = {
  origin: 'https://planets.com/',
  pathname: '/neptune/'
}

await unified()
  ...
```

> I'd argue there's no reasonable use case for treating a URL with an authority in it as "relative" for the purposes of this library.

Yeah, agreed. Interested in making a PR?

**Comment by iliana** [comment]
> Yeah, agreed. Interested in making a PR?

Yeah, can do that. May be a few days.

> This sounds a bit “adamant”. Why so certain?

Sorry it came off that way, I tend to be too terse. I didn't think about _requiring_ the base URL to be set, I sort of assumed it'd be optional; although I'm sure there's current consumers for whom adding that wouldn't make sense for their use case.

My plan is to send a PR that just checks for `//` at the start of the URL and treats it the same as if the `is-absolute-url` library returns `true`; it doesn't change too much and adds a mitigation.

**Comment by wooorm** [comment]
> Sorry it came off that way

No worries!

> although I'm sure there's current consumers for whom adding that wouldn't make sense for their use case.

One alternative, that might work: default to an origin of `localhost:8000`. It would still classify URLs to the *actual* page as external, but well, that’s the current state.
And I don’t imagine wrongly *not* marking `localhost:8000` as external is a (security) problem?

> it doesn't change too much and adds a mitigation.

Appreciate it!

**Comment by wooorm** [comment]
@iliana Still interested in/available to work on this? Can I help you?

**Comment by wooorm** [comment]
thanks iliana, released in [2.0.1](https://github.com/rehypejs/rehype-external-links/releases/tag/2.0.1)!
